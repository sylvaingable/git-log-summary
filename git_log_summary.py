from __future__ import annotations

import argparse
import csv
import re
import sys
from collections import defaultdict
from functools import reduce
from itertools import chain, cycle, tee
from typing import Generator, NamedTuple, TextIO

CHRONOLOGICAL_ORDERING = "chronological"
TOP_CHANGES_ORDERING = "top-changes"
TOP_COMMITS_ORDERING = "top-commits"

TEXT_OUTPUT = "text"
CSV_OUTPUT = "csv"


def main(args: list[str]):
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--ordering",
        choices=[CHRONOLOGICAL_ORDERING, TOP_CHANGES_ORDERING, TOP_COMMITS_ORDERING],
        default=CHRONOLOGICAL_ORDERING,
    )
    parser.add_argument(
        "--output-format", choices=[TEXT_OUTPUT, CSV_OUTPUT], default=TEXT_OUTPUT
    )
    parser.add_argument(
        "-x",
        "--exclude",
        nargs="?",
        action="append",
        type=str,
        default=[],
        metavar="author",
        dest="excluded_authors",
    )
    args = parser.parse_args(args)

    commits_logs = chunk_git_log(sys.stdin)
    commits = (parse_commit_log(commit_log) for commit_log in commits_logs)

    data = defaultdict(lambda: defaultdict(lambda: {"commits": 0, "changes": 0}))
    excluded_authors = set(args.excluded_authors)
    for commit in commits:
        if commit.author in excluded_authors or commit.email in excluded_authors:
            continue

        data[commit.date][commit.author]["commits"] += 1
        data[commit.date][commit.author]["changes"] += commit.changes

    for authors in data.values():
        authors_count = len(authors)
        authors["Total"]["commits"] = sum(a["commits"] for a in authors.values())
        authors["Total"]["changes"] = sum(a["changes"] for a in authors.values())
        authors["Average"]["commits"] = authors["Total"]["commits"] // authors_count
        authors["Average"]["changes"] = authors["Total"]["changes"] // authors_count

    commits_stats = sort_commits_stats(data, args.ordering)
    write_to_stdout(commits_stats, args.output_format)


def chunk_git_log(file_object: TextIO) -> Generator[str, None, None]:
    """Chunk a git log by commit."""
    chunk = []
    iterator = iter(file_object)
    while True:
        try:
            line = next(iterator)
        except StopIteration:
            if chunk:
                yield "".join(chunk)
            break
        if line.startswith("commit") and chunk:
            yield "".join(chunk)
            chunk = []
        if bool(line.strip()):  # Skip empty lines
            chunk.append(line)


class Commit(NamedTuple):
    author: str
    email: str
    date: str
    changes: int


# Regular expressions to match author, date and changes
author_re = re.compile(r"Author: (.+) <(.+)>")
date_re = re.compile(r"Date:\s+(\d{4}-\d{2})")
insertions_re = re.compile(r"(\d+) insertions?")
deletions_re = re.compile(r"(\d+) deletions?")
first_group = lambda match: next(iter(match.groups()))  # noqa: E731
first_group = lambda match: match.groups()[0]  # noqa: E731
second_group = lambda match: match.groups()[1]  # noqa: E731


def parse_commit_log(commit_log: str) -> Commit:
    author_match = author_re.search(commit_log)
    date_match = date_re.search(commit_log)
    insertions_match = insertions_re.search(commit_log)
    deletions_match = deletions_re.search(commit_log)

    author = first_group(author_match)
    email = second_group(author_match)
    commit_date = first_group(date_match)
    insertions = int(first_group(insertions_match)) if insertions_match else 0
    deletions = int(first_group(deletions_match)) if deletions_match else 0
    return Commit(author, email, commit_date, insertions + deletions)


def sort_commits_stats(
    commits_stats: dict[str, dict[str, dict[str, int]]], ordering: str
) -> list[str, dict[str, dict[str, int]]]:
    if ordering == CHRONOLOGICAL_ORDERING:
        return sorted(commits_stats.items(), key=lambda t: t[0])
    elif ordering == TOP_CHANGES_ORDERING:
        return sorted(
            commits_stats.items(),
            key=lambda t: t[1]["Average"]["changes"],
            reverse=True,
        )
    elif ordering == TOP_COMMITS_ORDERING:
        return sorted(
            commits_stats.items(),
            key=lambda t: t[1]["Average"]["commits"],
            reverse=True,
        )
    else:
        raise ValueError(f"Unknown ordering: {ordering}")


def to_text(
    file_object: TextIO, commits_stats: list[str, dict[str, dict[str, int]]]
) -> None:
    for date, authors in commits_stats:
        file_object.write(f"Date: {date}\n")
        for author, stats in authors.items():
            file_object.write(
                f'  {author:<20} commits: {stats["commits"]:>4}\tchanges: {stats["changes"]:>6}\n'
            )
        file_object.write("\n")


def to_csv(
    file_object: TextIO, commits_stats: list[str, dict[str, dict[str, int]]]
) -> None:
    # Find the longest list of authors
    authors = reduce(
        lambda acc, authors: authors if len(authors) > len(acc) else acc,
        (authors for _, authors in commits_stats),
        [],
    )
    # Each author requires two columns: one for commits and one for changes
    duplicated_authors = chain.from_iterable(zip(*tee(authors, 2)))
    authors_fields = tuple(
        f"{author} {col}"
        for author, col in zip(duplicated_authors, cycle(["commits", "changes"]))
    )
    # Use a dict writer to set empty values for authors that don't have commits on a
    # given date
    csv_writer = csv.DictWriter(file_object, fieldnames=["Date", *authors_fields])
    csv_writer.writeheader()
    for date, authors in commits_stats:
        row = {"Date": date}
        for author, stats in authors.items():
            row[f"{author} commits"] = stats["commits"]
            row[f"{author} changes"] = stats["changes"]
        csv_writer.writerow(row)


def write_to_stdout(
    commits_stats: list[str, dict[str, dict[str, int]]], output_format: str
):
    if output_format == TEXT_OUTPUT:
        to_text(sys.stdout, commits_stats)
    elif output_format == CSV_OUTPUT:
        to_csv(sys.stdout, commits_stats)


if __name__ == "__main__":
    main(sys.argv[1:])
