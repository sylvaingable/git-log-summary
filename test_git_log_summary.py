from io import StringIO

import pytest

from git_log_summary import main


@pytest.fixture
def mock_stdin(monkeypatch):
    git_log = StringIO(GIT_LOG_SAMPLE)
    monkeypatch.setattr("sys.stdin", git_log)


def test_text_summary(mock_stdin, capsys, snapshot):
    """
    Given the git log output of a project with 10 commits from 3 authors.
    When summarizing the git log using the default options.
    Then the text summary is correct.
    """
    main([])
    captured = capsys.readouterr()
    stdout = [line.split() for line in captured.out.rstrip().splitlines()]
    assert stdout == snapshot


def test_can_exclude_commits_by_author(mock_stdin, capsys):
    """
    Given the git log output of a project with 10 commits from 3 authors.
    When summarizing the git log excluding Alice.
    Then the text summary is correct.
    """
    main(["--exclude", "Alice"])
    captured = capsys.readouterr()
    stdout = captured.out.rstrip().splitlines()
    assert all("Alice" not in line for line in stdout)


GIT_LOG_SAMPLE = """
commit b7fa8da016c6e7570be246861ed8c2ca3f0b3abe
Author: Alice <alice@example.com>
Date:   2024-04-01

    more ignored words 

 7 files changed, 95 insertions(+), 107 deletions(-)

commit 23c48e63babb0cd39a847dac5458c67bea5aeabb
Author: Bob <bob@example.com>
Date:   2024-04-01

    Don't mess with Voodoo

 3 files changed, 53 insertions(+), 12 deletions(-)

commit fb0446c77f4eb45e7cae648ea77aa6da0a6c88a1
Author: Alice <alice@example.com>
Date:   2024-04-01

    should work now.

 15 files changed, 345 insertions(+), 437 deletions(-)

commit 179797fb36a6252426ddcdc02f5f10bd68b17e9a
Author: Charlie <charlie@example.com>
Date:   2024-04-01

    Temporary commit.

 8 files changed, 191 insertions(+), 11 deletions(-)

commit fe50e85b063036614a86fb1147a7315b43932f80
Author: Bob <bob@example.com>
Date:   2024-04-01

    pay no attention to the man behind the curtain

 2 files changed, 61 insertions(+), 14 deletions(-)

commit c63bfcf27050117d60e5edc43fa8a1667344c07b
Author: Bob <bob@example.com>
Date:   2024-04-01

    hmmm

 1 file changed, 31 deletions(-)

commit 32d0414004a0263cbec83469ff0fc4b3dcde2695
Author: Bob <bob@example.com>
Date:   2024-03-29

    We Had To Use Dark Magic To Make This Work

 1 file changed, 12 insertions(+)

commit 3ff3b1d296089003a93c75b736974cc33d3d4af5
Author: Bob <bob@example.com>
Date:   2024-03-28

    It's getting hard to keep up with the crap I've trashed

 18 files changed, 433 insertions(+), 275 deletions(-)

commit f18a416618d472aed9e9483609664099be392e50
Author: Charlie <charlie@example.com>
Date:   2024-03-28

    Gotta make you understand

 10 files changed, 238 insertions(+), 1 deletion(-)

commit 7a4db0e1cc4810207b2f49ef0006e4d7e0153c2c
Author: Alice <alice@example.com>
Date:   2024-03-28

    Spinning up the hamster...

 15 files changed, 577 insertions(+), 106 deletions(-)
"""
