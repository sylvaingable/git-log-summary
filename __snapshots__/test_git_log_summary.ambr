# serializer version: 1
# name: test_text_summary
  list([
    list([
      'Date:',
      '2024-03',
    ]),
    list([
      'Bob',
      'commits:',
      '2',
      'changes:',
      '720',
    ]),
    list([
      'Charlie',
      'commits:',
      '1',
      'changes:',
      '239',
    ]),
    list([
      'Alice',
      'commits:',
      '1',
      'changes:',
      '683',
    ]),
    list([
      'Total',
      'commits:',
      '4',
      'changes:',
      '1642',
    ]),
    list([
      'Average',
      'commits:',
      '1',
      'changes:',
      '547',
    ]),
    list([
    ]),
    list([
      'Date:',
      '2024-04',
    ]),
    list([
      'Alice',
      'commits:',
      '2',
      'changes:',
      '984',
    ]),
    list([
      'Bob',
      'commits:',
      '3',
      'changes:',
      '171',
    ]),
    list([
      'Charlie',
      'commits:',
      '1',
      'changes:',
      '202',
    ]),
    list([
      'Total',
      'commits:',
      '6',
      'changes:',
      '1357',
    ]),
    list([
      'Average',
      'commits:',
      '2',
      'changes:',
      '452',
    ]),
  ])
# ---
